### ISSUE TITLE
enhance iree-benchmark-module to provide required input sizes

### ISSUE THREAD

**Issue by powderluv** [iree-org/iree#12608]
### Request description

MLIR files are sometimes generated based on user input (image size,  base model channels etc) and we may not have access to the source MLIR files (since they are in the order of gigabytes). Have a way to benchmark the vmfb in such a way the required shapes are provided. 

In the real world we can evaluate and benchmark a end user system without asking for an MLIR dump by just running `iree-benchmark-module`  on their generated .vmfb

### What component(s) does this issue relate to?

Runtime

**Comment by powderluv** [comment]
Also capturing @benvanik response from Discord https://discord.com/channels/689900678990135345/689900680009482386/1084708941046480966
```
should never have to guess - the inputs are what you told them to be :)
if you want to automate things like that it's best to do it from the frontends - once we get down into the compiler there's a lot of stuff that gets hard to infer once we've lost the original information (and things like dynamic shapes will never work)
could have a pass like we do for the executable benchmarks that takes the frontend info and generates the shims but it'll always be untrustworthy
good for debugging, but I'd not run any benchmarks with it
(one example of frontend automation would be a function shim that dumps the inputs to a single .npy file you then pass in with --input=file.npy) 

```

**Comment by powderluv** [comment]
So maybe we squirrel away the input.npz in way and have a wrapper script around `iree-benchmark-module` to unzip the npz and pass them as inputs

**Comment by benvanik** [comment]
This isn't something we can support as it has too many cases that make it invalid for generalized benchmarking (no stateful models/training, no dynamic shapes, any input modeling a gather/lookup will be wrong/crash/have completely incorrect performance characteristics, etc). It could be a useful debugging feature in some cases but is not worth the complexity given how much configurability we already have. I don't trust benchmarks using this approach and don't really want to route people down this path - especially as we want to focus on dynamically shaped models.

All is not lost, though. Stepping back: someone had to have the source files at some point in order to be able to build the binary files and you don't need the _full_ source file in order to get the information to do what you're asking. Once someone has the source file they can run an iree-opt pass with elide-attrs set small to produce just the MLIR ops in mhlo/tosa/etc - which is then useful for performance analysis/triage/etc as well as having most of what you need. That could be a script that packaged things up and built a benchmark archive. Or you use `iree-compile --compile-to=input` (or some other phase) with elide-attrs set as that could be done via the API if using that for compilation. A useful pass could be one that just dumps the interface of the module (delete all private symbols and leave the public functions, remove the bodies, etc - effectively like generating a .h) and that could be used to slice out privileged stuff. Once you have that IR you can do whatever you want - like a pass that wraps functions with default inputs (like our existing executable benchmark stuff does) - or some python stuff to go walk the public functions and dump out a flagfile you can use to benchmark it. What you can hope to do with this stuff is limited, though, as you're only seeing half of the execution: the thing on the receiving end of of a call. This won't support stateful/training models and won't support dynamic shapes. If you're going to be passing in all zeros it can also skew performance depending on what the model is doing as for example every thread reading from location 0,0,0 it got from a gather vs over the entire 200MB range can have massive differences.

The solution that does support those things is going to the source of the call: save the inputs off when you have them in python/whatever other frontend. I'm not sure benchmarking things you can't run is a good idea, so then there's almost always going to be something that runs things you can shim into. If you produce a npy file with multiple arrays then you can pass that in with a single `--input=file.npy` and it'll have all the inputs. That isn't great if the inputs are very large but you could just as easily print out a flagfile with the input shapes to simulate splats. A flagfile is just one flag per line that you'd reference with `--flagfile=flags.txt`, and you could have something like:
```
--input=4xf32
--input=8xf32
etc
```
Walking over the args to an invocation function and dumping that to stdout or something in python should be pretty easy if you've already got code to run the model. The benefit here is that if there are dynamic shapes you'll have captured a representative example - if you were just looking at what the compiler gets you'd never be able to know what that input was expected to be. You could also print out many calls and then run your benchmarks over all of them to see what different sizes do. Going further you can print out a trace yaml file for `iree-benchmark-trace` which supports stateful models/training but it's the same concept: python layer dumps out a text file and runtime tool takes that text file in.

Just saw your comment: that's a zip/tar file and a shell script/command line/python script :) If you just want unzip + call iree-benchmark-module --module=foo --input=bar (or --flagfile=bar) then it's simpler to do that. That's also better than doing it at once as unless you only plan on running the benchmark one time you've now done the expensive thing (unarchiving) once for as many runs as you need.

There's lots of solutions in this space - you could dump flagfiles from iree-compile, for example - but that inherits a lot of the issues with the compile-time approach. Things get complex fast and I'm not sure it's better to tightly couple this stuff than it is to just use the normal tools and tailor it to your situation. It could be useful to others working with similar types of models and similar restrictions and if they understood the caveats then it could be under scripts/ (like compile-benchmark-package.py/run-benchmark-package.py or something). iree-benchmark-module/iree-benchmark-trace (which is even more configurable) and npy/flagfiles/etc should be what's needed to build such workflows, but if there's python shims/examples/etc those would be useful to have. Would be nice to have one-liners/mixins for "monkey patch tf session run to dump all the input shapes to a file" - that's the most interesting bit and everything after that is mostly the same (pass in inputs/flags/etc).

TLDR if you're automating this you're going to have some script running iree-benchmark-module and it can run 1 other command first if you want to archive things. If you're running things by hand then flagfiles/npy files should be sufficient (you're passing --device= and --module=, so one extra --flagfile=/--input= is fine). You can dump command lines, flagfiles, etc from your frontend.

**Comment by allieculp** [comment]
@powderluv Do you want to leave this open?
